src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -1184,6 +1203,205 @@ void HNSWIndex<DataType, DistType>::resizeIndexInternal(size_t new_max_elements)
+
+    // Perform mutual updates: go over the node's neighbors and overwrite the neighbors to remove
+    // that are still exist.
+    size_t node_neighbors_idx = 0;
+    for (size_t i = 0; i < node_neighbors_count; i++) {
+        if (!std::binary_search(nodes_to_update.begin(), nodes_to_update.end(),
+                                node_neighbors[i])) {
+            // The repaired node added a new neighbor that we didn't account for before in the
+            // meantime - leave it as is.
+            node_neighbors[node_neighbors_idx++] = node_neighbors[i];
+            continue;
+        }
+        auto it = std::find(chosen_neighbors.begin(), chosen_neighbors.end(), node_neighbors[i]);
+        if (it != chosen_neighbors.end()) {
+            // A chosen neighbor is already connected to the node - leave it as is.
+            node_neighbors[node_neighbors_idx++] = node_neighbors[i];
+            // Swap the last element with the current one (equivalent to removing the neighbor from
+            // the list).
+            *it = *(chosen_neighbors.end() - 1);
+            chosen_neighbors.pop_back();
+            continue;
+        }
+        // Now we know that we are looking at a neighbor that needs to be removed.
+        auto removed_node = node_neighbors[i];
+        auto *removed_node_incoming_edges = getIncomingEdgesPtr(removed_node, level);
+        // Perform the mutual update:
+        // if the removed node id (the node's neighbour to be removed)
+        // wasn't pointing to the node (i.e., the edge was uni-directional),
+        // we should remove the current neighbor from the node's incoming edges.
+        // otherwise, the edge turned from bidirectional to uni-directional, so we insert it to the
+        // neighbour's incoming edges set. Note: we assume that every update is performed atomically
+        // mutually, so it should be sufficient to look at the removed node's incoming edges set
+        // alone.
+        auto incoming_edges_it = std::find(removed_node_incoming_edges->begin(),
+                                           removed_node_incoming_edges->end(), node_id);
+        if (incoming_edges_it != removed_node_incoming_edges->end()) {
+            // Swap the last element with the current one (equivalent to removing the neighbor from
+            // the list).
+            *incoming_edges_it = *(removed_node_incoming_edges->end() - 1);
+            removed_node_incoming_edges->pop_back();
+        } else {
+            node_incoming_edges->push_back(removed_node);
+        }
+    }
+
+    // Go over the chosen new neighbors that are not connected yet and perform updates.
+    for (auto chosen_id : chosen_neighbors) {
+        if (node_neighbors_idx == max_M_cur) {
+            // Cannot add more new neighbors, we reached the capacity.
+            // TODO: move this message to a log.
+            std::cout << "couldn't add all the chosen new nodes upon updating " << node_id
```

Comment:
It should go to Redis log anyway and then we can decide the log level. But in general, I think that this information should actually be printed NOT just in debug because it should happen in rare conditions, and it's an indication of heavy updates situation that may decrease the index quality - I believe that this is a useful "supportability" information...

Reply:
yep that's the plan (to enable logging BEFORE we merge this feature branch)